Add ResourceReference to task spec

liron-l opened the pull request:
Add missing `ResourceReference` as part of the swarmapi conversion

This is required for populating the secret reference to swarm plugins.

docker/swarmkit#2380

Signed-off-by: Liron Levin <liron@twistlock.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

Commit:
Add ResourceReference to task spec

Add missing `ResourceReference` as part of the swarmapi conversion

This is required for populating the secret reference to swarm plugins.

docker/swarmking#2380

Signed-off-by: Liron Levin <liron@twistlock.com>

liron-l commented:
@aaronlehmann, @riyazdf PTAL.

aaronlehmann commented on api/types/swarm/task.go:
```diff
@@ -116,6 +117,24 @@ type ResourceRequirements struct {
+	ResourceTypeSecret ResourceType = 1
+	// ResourceTypeTask is the config resource type
+	ResourceTypeConfig ResourceType = 2
+)
```
In the JSON API, I think these should be strings rather than integers.

thaJeztah commented:
CI is failing;

```
11:40:03 api/types/swarm/task.go:128:2:warning: comment on exported const ResourceTypeConfig should be of the form "ResourceTypeConfig ..." (golint)
11:40:03 api/types/swarm/task.go:126:2:warning: comment on exported const ResourceTypeSecret should be of the form "ResourceTypeSecret ..." (golint)
```

Also, don't forget to update the swagger.yml with changes, and the API history;

- https://github.com/moby/moby/blob/master/api/swagger.yaml
- https://github.com/moby/moby/blob/master/docs/api/version-history.md

liron-l commented:
Thanks @aaronlehmann, I've fixed the code according to your comments. 
@thaJeztah I've modified the swagger API and API history, let me know if additional changes are required.

cpuguy83 reviewed (commented):
This seems ok.
Wondering if these are actually wired up anywhere. If not I'd prefer not to document the API.

Also wondering why/when a user would request a task reference.

cpuguy83 commented:
> more work work is required in order to populate the referenced secrets to the plugin service

Also probably the container controller should handle this as well.

liron-l commented:
Thanks @cpuguy83. I removed the API reference (and I assume you mean plugin controller).  

I think that in order to add secret injection support, we need to do the following:
1. Inject secrets plugin backed (simillar to container plugin that uses the swarm agent backend)

2. If we can assume that secrets are always mounted with root permissions we don't need additional metadata. However, if not, we should add `SecretReference` to the `PluginSpec` (similar to `ContainerSpec`)

3. In swarmkit, inject the additional secret resource references in  [taskRestrictedSecretProvider](https://github.com/docker/swarmkit/blob/master/agent/secrets/secrets.go#L80), otherwise they will not be populated.

Please let me know if the above makes sense.

cpuguy83 commented:
> If we can assume that secrets are always mounted with root permissions we don't need additional metadata. However, if not, we should add SecretReference to the PluginSpec (similar to ContainerSpec)

Yes we should add a secret reference to `PluginSpec` so this can be configured.

cpuguy83 commented:
@liron-l What's the status of this one?

sirlatrom commented:
@anshulpundir Any chance I could ask you to consider this PR? I'll be happy to carry it forward with any required changes.

olljanat commented:
Derek add label: rebuild/janky

cpuguy83 commented:
@sirlatrom So there are two aspects here.

1. We really wanted to use this as a way to allow for pulling images using a secret without having to forward the secret from the client itself as is required today with `--with-registry-auth`
2. Would be nice if plugins could support secrets/configs.

I think 1 is a much more pertinent need and affects both the swarm container runtime and the generic runtime.

For 2 some of this secret/config handling will need to be extracted out into something more consumable. But I'm also not sure 2 has really been requested as of yet. This probably doesn't need to be wired up to start with.

sirlatrom commented:
> @sirlatrom So there are two aspects here.
> 
>     1. We really wanted to use this as a way to allow for pulling images using a secret without having to forward the secret from the client itself as is required today with `--with-registry-auth`

You're absolutely right, of course. Is there someone working actively on that, or should I give it a try?

cpuguy83 commented:
@sirlatrom The only bits that were worked on are what you see here and the linked swarmkit repo... of course neither are merged.

thaJeztah commented:
With https://github.com/docker/swarmkit/pull/2380 not yet being merged, should we close this one?

aaronlehmann commented:
Sounds like we should.
